// fnbr/dist/src/structures/party/ClientPartyMemberMeta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const defaultPartyMemberMeta_json_1 = tslib_1.__importDefault(require("../../../resources/defaultPartyMemberMeta.json"));
const Util_1 = require("../../util/Util");
const PartyMemberMeta_1 = tslib_1.__importDefault(require("./PartyMemberMeta"));
/**
 * Represents the client's party member meta
 */
class ClientPartyMemberMeta extends PartyMemberMeta_1.default {
    /**
     * @param member The party member
     * @param schema The schema
     */
    constructor(member, schema) {
        super({ ...defaultPartyMemberMeta_json_1.default });
        this.member = member;
        const defaultCharacter = (0, Util_1.getRandomDefaultCharacter)();
        this.update({
            'Default:AthenaCosmeticLoadout_j': JSON.stringify({
                AthenaCosmeticLoadout: {
                    ...JSON.parse(defaultPartyMemberMeta_json_1.default['Default:AthenaCosmeticLoadout_j']).AthenaCosmeticLoadout,
                    characterPrimaryAssetId: `AthenaCharacter:${defaultCharacter}`,
                },
            }),
            'Default:CampaignHero_j': JSON.stringify({
                CampaignHero: {
                    heroItemInstanceId: '',
                    heroType: `/Game/Athena/Heroes/${defaultCharacter.replace('CID', 'HID')}.${defaultCharacter.replace('CID', 'HID')}`,
                },
            }),
            'Default:PlatformData_j': JSON.stringify({
                PlatformData: {
                    platform: {
                        platformDescription: {
                            name: member.client.config.platform,
                            platformType: 'DESKTOP',
                            onlineSubsystem: 'None',
                            sessionType: '',
                            externalAccountType: '',
                            crossplayPool: 'DESKTOP',
                        },
                    },
                    uniqueId: 'INVALID',
                    sessionId: '',
                },
            }),
        }, true);
        if (schema)
            this.update(schema, true);
    }
}
exports.default = ClientPartyMemberMeta;
