// fnbr/dist/src/structures/ClientPartyMeta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const defaultPartyMeta_json_1 = tslib_1.__importDefault(require("../../resources/defaultPartyMeta.json"));
const PartyMeta_1 = tslib_1.__importDefault(require("./PartyMeta"));
/**
 * Represents the client's party meta
 */
class ClientPartyMeta extends PartyMeta_1.default {
    /**
     * @param party The party
     * @param schema The schema
     */
    constructor(party, schema) {
        super({ ...defaultPartyMeta_json_1.default });
        this.party = party;
        this.refreshSquadAssignments();
        this.updatePrivacy();
        if (schema)
            this.update(schema, true);
    }
    /**
     * Refreshes the member positions
     */
    refreshSquadAssignments() {
        var _a;
        let i = 0;
        const assignments = [];
        if (this.party.me && !this.party.hiddenMemberIds.has(this.party.me.id)) {
            assignments.push({
                memberId: (_a = this.party.client.user) === null || _a === void 0 ? void 0 : _a.id,
                absoluteMemberIdx: 0,
            });
            i += 1;
        }
        this.party.members.forEach((m) => {
            var _a;
            if (m.id !== ((_a = this.party.client.user) === null || _a === void 0 ? void 0 : _a.id) && !this.party.hiddenMemberIds.has(m.id)) {
                assignments.push({
                    memberId: m.id,
                    absoluteMemberIdx: i,
                });
                i += 1;
            }
        });
        return this.set('Default:RawSquadAssignments_j', {
            RawSquadAssignments: assignments,
        });
    }
    updatePrivacy() {
        this.set('Default:PrivacySettings_j', {
            PrivacySettings: {
                partyType: this.party.config.privacy.partyType,
                partyInviteRestriction: this.party.config.privacy.inviteRestriction,
                bOnlyLeaderFriendsCanJoin: this.party.config.privacy.onlyLeaderFriendsCanJoin,
            },
        });
    }
}
exports.default = ClientPartyMeta;
