// fnbr/dist/src/structures/EpicgamesServerStatusIncident.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Base_1 = tslib_1.__importDefault(require("../Base"));
/**
 * Represents an Epicgames server status
 */
class EpicgamesServerStatusIncident extends Base_1.default {
    /**
     * @param client The main client
     * @param data The server status data
     */
    constructor(client, data) {
        super(client);
        this.id = data.id;
        this.pageId = data.page_id;
        this.name = data.name;
        this.status = data.status;
        this.impact = data.impact;
        this.shortLink = data.short_link;
        this.createdAt = new Date(data.created_at);
        this.updatedAt = data.updated_at ? new Date(data.updated_at) : undefined;
        this.monitoringAt = data.monitoring_at ? new Date(data.monitoring_at) : undefined;
        this.resolvedAt = data.resolved_at ? new Date(data.resolved_at) : undefined;
        this.incidentUpdates = data.incident_updates.map((u) => ({
            id: u.id,
            status: u.status,
            body: u.body,
            createdAt: new Date(u.created_at),
            displayAt: new Date(u.display_at),
            updatedAt: new Date(u.updated_at),
        }));
    }
}
exports.default = EpicgamesServerStatusIncident;
