// fnbr/dist/src/structures/NewsMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Base_1 = tslib_1.__importDefault(require("../Base"));
const Image_1 = tslib_1.__importDefault(require("./Image"));
const NewsMessageVideo_1 = tslib_1.__importDefault(require("./NewsMessageVideo"));
/**
 * Represents a fortnite news message
 */
class NewsMessage extends Base_1.default {
    /**
     * @param client The main client
     * @param data The news message data
     */
    constructor(client, data) {
        var _a;
        super(client);
        const newsData = data.contentFields;
        this.title = newsData.title;
        this.body = newsData.body;
        this.images = newsData.image.map((i) => new Image_1.default(this.client, i));
        this.entryType = newsData.entryType;
        this.buttonTextOverride = newsData.buttonTextOverride;
        this.tabTitleOverride = newsData.tabTitleOverride;
        this.tileImage = new Image_1.default(this.client, (_a = newsData.tileImage) === null || _a === void 0 ? void 0 : _a[0]);
        this.playlist = newsData.playlistId ? {
            id: newsData.playlistId,
        } : undefined;
        this.offer = newsData.offerId ? {
            id: newsData.offerId,
            action: newsData.offerAction,
        } : undefined;
        this.video = newsData.videoUID ? new NewsMessageVideo_1.default(this.client, {
            videoAutoplay: newsData.videoAutoplay,
            videoFullscreen: newsData.videoFullscreen,
            videoLoop: newsData.videoLoop,
            videoMute: newsData.videoMute,
            videoStreamingEnabled: newsData.videoStreamingEnabled,
            videoUID: newsData.videoUID,
            videoVideoString: newsData.videoVideoString,
        }) : undefined;
    }
}
exports.default = NewsMessage;
